Short circuit Clamp when value = min

Buildstarted opened the pull request:
When `value == min` we can return immediately and not execute the next if statement.

Benchmarks
============
```
|                 Method | value | min | max |      Mean |     Error |    StdDev |
|----------------------- |------ |---- |---- |----------:|----------:|----------:|
|              MathClamp |     0 |   0 |  10 | 0.5672 ns | 0.0073 ns | 0.0068 ns |
|  ShortCircuitMathClamp |     0 |   0 |  10 | 0.2581 ns | 0.0090 ns | 0.0084 ns |
```

Seems to be approximately twice as fast when `value == min`.

I did not alter double, decimal, or float.

Buildstarted committed:
Short circuit Clamp when value = min

When value == min it makes sense to not run the next if statement and
just return a value immediately.

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/corefx?pullRequest=40855) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/corefx?pullRequest=40855) before we can accept your contribution.<br/><br/>:x: Buildstarted [sign now](https://cla.dotnetfoundation.org/dotnet/corefx?pullRequest=40855)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/corefx?pullRequest=40855) it.</sub>

danmosemsft commented:
What about `else if (value > max)` ?

Buildstarted commented:
if `value == min` then `if (value < min)` would fail which would necessitate the `else if (value > max)` branch to be tested.

tannergooding commented:
Looks reasonable to me, the codegen should basically be the same but with a `jle` rather than a `jl` instruction which is probably more beneficial for what the JIT is currently generating:
```diff
     L0000: cmp ecx, edx
-    L0002: jae L0007      
+    L0002: ja L0007      
     L0004: mov eax, edx
     L0006: ret
     L0007: cmp ecx, r8d
     L000a: jbe L0010
     L000c: mov eax, r8d
     L000f: ret
     L0010: mov eax, ecx
     L0012: ret
```

You might see a tiny increase if you also do `else if (value >= max)`.

However, this is probably a case where the JIT emitting a  `cmov` would be more beneficial since the correct path isn't easily predicted. The better codegen (and what MSVC, GCC, and Clang generate), is something like:
```asm
; ecx - value
; edx - min
; r8d - max

mov eax, r8d
cmp r8d, ecx
cmovae eax, ecx
cmp ecx, edx
cmovb eax, edx
ret
```

Buildstarted commented:
```
|               Method | value | min | max |      Mean |     Error |    StdDev |    Median |
|--------------------- |------ |---- |---- |----------:|----------:|----------:|----------:|
|            MathClamp |     0 |   0 |  10 | 0.5417 ns | 0.0165 ns | 0.0154 ns | 0.5365 ns |
|  MathClampWithMaxGTE |     0 |   0 |  10 | 0.5954 ns | 0.0172 ns | 0.0161 ns | 0.5903 ns |
|            MathClamp |     5 |   0 |  10 | 0.5845 ns | 0.0221 ns | 0.0196 ns | 0.5829 ns |
|  MathClampWithMaxGTE |     5 |   0 |  10 | 0.5807 ns | 0.0232 ns | 0.0205 ns | 0.5755 ns |
|            MathClamp |    10 |   0 |  10 | 0.5794 ns | 0.0171 ns | 0.0160 ns | 0.5741 ns |
|  MathClampWithMaxGTE |    10 |   0 |  10 | 0.5670 ns | 0.0108 ns | 0.0096 ns | 0.5665 ns |
```
Ran some numbers with `value >= max` it does look like there's a minor increase in speed when `value >= max` except I'm not quite sure what to make of the first result. I ran it a few times and got similar results each time.

Buildstarted committed:
Short circuit value >= max

There seems to be a minor increase in speed if we short circuit max as
well as we don't have to jump.

Buildstarted commented:
Can someone help me with the failing builds? I ran the build command that builds, buildstests and runs the tests and it looks good.

Wraith2 commented:
I don't think they're related to what you're doing. The log of what actually failed is in the helix upload step but you have to scroll right to see it, https://helix.dot.net/api/2019-06-17/jobs/c0510143-bf05-4c93-8f90-b95c1478eb52/workitems/System.Net.Security.Tests/console

jkotas commented:
Changes under src/Common/src/CoreLib/System/Math.cs  should be submitted via https://github.com/dotnet/coreclr repo. These files are not actually built in CoreFX.

jkotas commented:
> correct path isn't easily predicted

Yep, it means that this change makes Clamp faster for some distribution of inputs, and slower for some distribution of inputs. Do we have an idea what is the typical distribution of inputs for Clamp?

For example, the following benchmark that calls Clamp on random input is close to 2x slower with this change:

```
using System;
using System.Diagnostics;

class Program
{
    static void Main(string[] args)
    {
         int[] values = new int[1000];
         Random r = new Random();
         for (int i = 0; i < values.Length; i++)
         {
             values[i] = r.Next(5) + 5;
         }

         int[] result = new int[values.Length];
         for (;;)
         {
              var sw = new Stopwatch();
              sw.Start();
              for (int j = 0; j < 1000000; j++)
                   for (int i = 0; i < values.Length; i++)
                       result[i] = Math.Clamp(values[i], 5, 9);
              Console.WriteLine(sw.Elapsed);
         }
    }
}
```

tannergooding commented:
> Do we have an idea what is the typical distribution of inputs for Clamp?

I don't think we have any concrete data on this today. The best solution is likely to just follow what the native compilers do and generate branchless code (which I would guess is done because there aren't any concrete metrics on what the typical distribution is).

Buildstarted commented:
Thanks for the insight, jkotas. I was looking at it from a pure logic pov but didn't have any idea how the compiler might compile it or if it would alter how the compiler works. I should have known smarter minds than I would know better. :)
